libcudacxx/include/cuda/std/__cuda/semaphore.h

```diff
@@ -23,23 +24,147 @@
 
 _LIBCUDACXX_BEGIN_NAMESPACE_CUDA
 
+void simple_backoff() {
+  int wait_iterations = 1;
+  volatile int ctr = 0;
+  const int max_iterations = 1024;
+
+  for (int i = 0; i < wait_iterations; ++i)
+  {
+    ctr++;
+  }
+  wait_iterations = (wait_iterations * 2 >= max_iterations) ? max_iterations : wait_iterations * 2;
+}
+
+
 template <thread_scope _Sco, ptrdiff_t __least_max_value = INT_MAX>
-class counting_semaphore : public _CUDA_VSTD::__semaphore_base<__least_max_value, _Sco>
+class __fair_semaphore
 {
-  static_assert(__least_max_value <= _CUDA_VSTD::__semaphore_base<__least_max_value, _Sco>::max(), "");
+private:
+  const ptrdiff_t max_;
+  cuda::atomic<ptrdiff_t, _Sco> current;
+  cuda::atomic<ptrdiff_t, _Sco> tickets;
 
 public:
-  _LIBCUDACXX_INLINE_VISIBILITY constexpr counting_semaphore(ptrdiff_t __count = 0)
-      : _CUDA_VSTD::__semaphore_base<__least_max_value, _Sco>(__count)
-  {}
-  ~counting_semaphore() = default;
+    static_assert(__least_max_value >= 0, "The least maximum value must be a positive number");
+
+    _LIBCUDACXX_INLINE_VISIBILITY bool try_acquire()
+  {
+    ptrdiff_t available_tickets = max_ - tickets.load();
+    return (available_tickets > 0) ? true : false;
+  }
+  
+  _LIBCUDACXX_INLINE_VISIBILITY bool our_turn(ptrdiff_t ticket_num)
+  {
+    ptrdiff_t serving_ticket = current.load();
+    return ticket_num == serving_ticket;
+  }
+
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr ptrdiff_t max() noexcept
+  {
+    return _CUDA_VSTD::numeric_limits<ptrdiff_t>::max();
+  }
+
+  // should block until it's their turn. Right? 
+  _LIBCUDACXX_INLINE_VISIBILITY void acquire()
+  { 
+    // if there aren't any tickets available we block
+    while(!try_acquire()) {
+      simple_backoff();
+    }
+    ptrdiff_t t = tickets.fetch_add(1, cuda::memory_order_relaxed);
```

Comment:
If the prior lock unblocks because there is 1 ticket available, but some other thread gets that ticket before this one, then this `fetch_add` will overflow the tickets.